src/relayclient/KnownRelaysManager.ts

```diff
@@ -103,14 +104,51 @@ export default class KnownRelaysManager implements IKnownRelaysManager {
     return origRelays.filter(this.relayFilter)
   }
 
+  splitRange (fromBlock: number, toBlock: number, splits: number): Array<{fromBlock: number, toBlock: number}> {
+    const totalBlocks = toBlock - fromBlock + 1
+    const splitSize = Math.ceil(totalBlocks / splits)
+
+    const ret: Array<{ fromBlock: number, toBlock: number }> = []
+    let b
+    for (b = fromBlock; b < toBlock; b += splitSize) {
+      ret.push({ fromBlock: b, toBlock: Math.min(toBlock, b + splitSize - 1) })
+    }
+    return ret
+  }
+
+  // return events from hub. split requested range into "window parts", to avoid
+  // fetching too many events at once.
+  async getPastEventsForHub (fromBlock: number, toBlock: number): Promise<EventData[]> {
+    let relayEventParts: any[]
+    while (true) {
+      const rangeParts = this.splitRange(fromBlock, toBlock, this.config.relayLookupWindowParts)
+      try {
+        // eslint-disable-next-line @typescript-eslint/promise-function-async
+        relayEventParts = await Promise.all(rangeParts.map(({ fromBlock, toBlock }) =>
+          this.contractInteractor.getPastEventsForHub([], {
+            fromBlock,
+            toBlock
+          })))
+        break
+      } catch (e) {
+        if (e.toString().match(/query returned more than/) != null &&
+          this.config.relayLookupWindowBlocks > this.config.relayLookupWindowParts
+        ) {
+          this.config.relayLookupWindowParts *= 2
+          continue
```

Comment:
It is not clear what is going on here. Also, there is no clear halting condition that does not rely on remote API response. I suggest if the value is not 1, don't try to paginate. If it is 1, maybe try also 5, 10 and halt there.  Say we need  to split in like 10 pages.
With this code it will try 2, 4, 8 before it reaches 16. I think if we can't make it in 10 pages, we bail.

Reply:
ok, maybe double is little. I still like binaries, so put 4, 16 
halting condition is needed, so I'll stop after that.

Reply:
also, if the client's default is (say) 3 and it fails on "too many results". why not attempt to raise it?